File: runtime/microvm/controller_routes.go
package microvm

import (
	"encoding/json"
	"errors"
	"strconv"
	"strings"

	apptheory "github.com/theory-cloud/apptheory/runtime"
)

// RegisterControllerRoutes registers the canonical M16 MicroVM controller HTTP routes on an AppTheory app.
func RegisterControllerRoutes(app *apptheory.App, controller *Controller) (*apptheory.App, error) {
	if app == nil {
		return nil, errors.New("apptheory: microvm controller route registration requires an app")
	}
	if controller == nil {
		return app, safeError(ErrorCodeControllerIncomplete, "apptheory: microvm controller route registration requires a controller", "")
	}
	routes := []struct {
		method  string
		path    string
		command Command
	}{
		{"POST", "/microvms", CommandRun},
		{"GET", "/microvms", CommandList},
		{"GET", "/microvms/{session_id}", CommandGet},
		{"POST", "/microvms/{session_id}/suspend", CommandSuspend},
		{"POST", "/microvms/{session_id}/resume", CommandResume},
		{"DELETE", "/microvms/{session_id}", CommandTerminate},
		{"POST", "/microvms/{session_id}/auth-token", CommandAuthToken},
		{"POST", "/microvms/{session_id}/shell-auth-token", CommandShellAuthToken},
		// Compatibility route for callers created during the earlier M16 contract correction.
		{"POST", "/microvms/{session_id}/shell-token", CommandShellAuthToken},
	}
	for _, route := range routes {
		if _, err := app.HandleStrict( //nolint:staticcheck // RegisterControllerRoutes preserves its error-returning API.
			route.method,
			route.path,
			controllerRouteHandler(controller, route.command),
			apptheory.RequireAuth(),
		); err != nil {
			return app, err
		}
	}
	for _, method := range controllerInvokeMethods() {
		for _, path := range []string{"/microvms/{session_id}/invoke", "/microvms/{session_id}/invoke/{proxy+}"} {
			if _, err := app.HandleStrict( //nolint:staticcheck // RegisterControllerRoutes preserves its error-returning API.
				method,
				path,
				controllerInvokeRouteHandler(controller),
				apptheory.RequireAuth(),
			); err != nil {
				return app, err
			}
		}
	}
	return app, nil
}

// RegisterMicroVMControllerRoutes is an explicit alias for RegisterControllerRoutes.
func RegisterMicroVMControllerRoutes(app *apptheory.App, controller *Controller) (*apptheory.App, error) {
	return RegisterControllerRoutes(app, controller)
}

type controllerRoutePayload struct {
	TenantID                    string              `json:"tenant_id"`
	Namespace                   string              `json:"namespace"`
	SessionID                   string              `json:"session_id"`
	ImageRef                    string              `json:"image_ref"`
	ImageVersion                string              `json:"image_version"`
	NetworkConnectorRef         string              `json:"network_connector_ref"`
	IngressNetworkConnectorRefs []string            `json:"ingress_network_connector_refs"`
	EgressNetworkConnectorRefs  []string            `json:"egress_network_connector_refs"`
	SessionSpec                 SessionSpec         `json:"session_spec"`
	IdlePolicy                  *ProviderIdlePolicy `json:"idle_policy"`
	MaximumDurationSeconds      int32               `json:"maximum_duration_seconds"`
	TTLSeconds                  int32               `json:"ttl_seconds"`
	AllowedPortScope            []ProviderPortScope `json:"allowed_port_scope"`
	MaxResults                  int32               `json:"max_results"`
}

func controllerRouteHandler(controller *Controller, command Command) apptheory.Handler {
	return func(ctx *apptheory.Context) (*apptheory.Response, error) {
		request, safe := controllerRequestFromHTTP(ctx, command)
		if safe.Code != "" {
			return controllerHTTPResponse(controllerErrorResponse(request, safe)), nil
		}
		response, err := controller.Handle(ctx.Context(), request)
		if err != nil && response.Error == nil {
			safe := asSafeError(err, request.RequestID)
			response = controllerErrorResponse(request, safe)
		}
		return controllerHTTPResponse(response), nil
	}
}

func controllerInvokeRouteHandler(controller *Controller) apptheory.Handler {
	return func(ctx *apptheory.Context) (*apptheory.Response, error) {
		request, safe := controllerInvokeRequestFromHTTP(ctx)
		if safe.Code != "" {
			return controllerHTTPResponse(controllerErrorResponse(ControllerRequest{
				Command:     CommandInvoke,
				RequestID:   request.RequestID,
				TenantID:    request.TenantID,
				Namespace:   request.Namespace,
				AuthContext: request.AuthContext,
				SessionID:   request.SessionID,
			}, safe)), nil
		}
		output, err := controller.Invoke(ctx.Context(), request)
		if err != nil {
			safe := asSafeError(err, request.RequestID)
			return controllerHTTPResponse(controllerErrorResponse(ControllerRequest{
				Command:     CommandInvoke,
				RequestID:   request.RequestID,
				TenantID:    request.TenantID,
				Namespace:   request.Namespace,
				AuthContext: request.AuthContext,
				SessionID:   request.SessionID,
			}, safe)), nil
		}
		return controllerInvokeHTTPResponse(output), nil
	}
}

func controllerRequestFromHTTP(ctx *apptheory.Context, command Command) (ControllerRequest, SafeError) {
	request := ControllerRequest{Command: normalizeCommand(command)}
	if ctx == nil {
		err := safeError(ErrorCodeInvalidControllerRequest, "apptheory: microvm controller route context is missing", "")
		return request, err
	}
	payload := controllerRoutePayload{}
	if len(ctx.Request.Body) > 0 {
		if err := json.Unmarshal(ctx.Request.Body, &payload); err != nil {
			request.RequestID = strings.TrimSpace(ctx.RequestID)
			err := safeError(ErrorCodeInvalidControllerRequest, "apptheory: microvm controller route request is malformed", request.RequestID)
			return request, err
		}
	}
	pathSessionID := strings.TrimSpace(ctx.Param("session_id"))
	bodySessionID := strings.TrimSpace(payload.SessionID)
	if pathSessionID != "" && bodySessionID != "" && pathSessionID != bodySessionID {
		request.RequestID = strings.TrimSpace(ctx.RequestID)
		err := safeError(ErrorCodeTenantBindingViolation, "apptheory: microvm controller route session binding mismatch", request.RequestID)
		return request, err
	}
	if ctx.TenantID != "" && strings.TrimSpace(payload.TenantID) != "" && strings.TrimSpace(payload.TenantID) != strings.TrimSpace(ctx.TenantID) {
		request.RequestID = strings.TrimSpace(ctx.RequestID)
		err := safeError(ErrorCodeTenantBindingViolation, "apptheory: microvm controller route tenant binding mismatch", request.RequestID)
		return request, err
	}
	if ctx.TenantID != "" && strings.TrimSpace(ctx.Query("tenant_id")) != "" && strings.TrimSpace(ctx.Query("tenant_id")) != strings.TrimSpace(ctx.TenantID) {
		request.RequestID = strings.TrimSpace(ctx.RequestID)
		err := safeError(ErrorCodeTenantBindingViolation, "apptheory: microvm controller route tenant binding mismatch", request.RequestID)
		return request, err
	}
	sessionID := defaultString(pathSessionID, bodySessionID)
	tenantID := defaultString(strings.TrimSpace(ctx.TenantID), defaultString(payload.TenantID, ctx.Query("tenant_id")))
	namespace := defaultString(strings.TrimSpace(payload.Namespace), defaultString(ctx.Header("x-namespace-id"), ctx.Query("namespace")))
	request = ControllerRequest{
		Command:                     normalizeCommand(command),
		RequestID:                   strings.TrimSpace(ctx.RequestID),
		TenantID:                    tenantID,
		Namespace:                   namespace,
		AuthContext:                 AuthContext{Subject: strings.TrimSpace(ctx.AuthIdentity), TenantID: strings.TrimSpace(ctx.TenantID), Namespace: namespace},
		SessionID:                   sessionID,
		ImageRef:                    payload.ImageRef,
		ImageVersion:                payload.ImageVersion,
		NetworkConnectorRef:         payload.NetworkConnectorRef,
		IngressNetworkConnectorRefs: append([]string(nil), payload.IngressNetworkConnectorRefs...),
		EgressNetworkConnectorRefs:  append([]string(nil), payload.EgressNetworkConnectorRefs...),
		SessionSpec:                 cloneSessionSpec(payload.SessionSpec),
		IdlePolicy:                  payload.IdlePolicy,
		MaximumDurationSeconds:      payload.MaximumDurationSeconds,
		TTLSeconds:                  payload.TTLSeconds,
		AllowedPortScope:            append([]ProviderPortScope(nil), payload.AllowedPortScope...),
		MaxResults:                  firstPositiveInt32(payload.MaxResults, parseInt32(ctx.Query("max_results"))),
	}
	request = normalizeControllerRequest(request)
	return request, SafeError{}
}

func controllerInvokeRequestFromHTTP(ctx *apptheory.Context) (ControllerInvokeRequest, SafeError) {
	request := ControllerInvokeRequest{}
	if ctx == nil {
		err := safeError(ErrorCodeInvalidControllerRequest, "apptheory: microvm controller route context is missing", "")
		return request, err
	}
	if ctx.TenantID != "" && strings.TrimSpace(ctx.Query("tenant_id")) != "" && strings.TrimSpace(ctx.Query("tenant_id")) != strings.TrimSpace(ctx.TenantID) {
		request.RequestID = strings.TrimSpace(ctx.RequestID)
		err := safeError(ErrorCodeTenantBindingViolation, "apptheory: microvm controller route tenant binding mismatch", request.RequestID)
		return request, err
	}
	sessionID := strings.TrimSpace(ctx.Param("session_id"))
	proxyPath := strings.TrimSpace(ctx.Param("proxy"))
	if proxyPath == "" {
		proxyPath = "/"
	}
	namespace := defaultString(ctx.Header("x-namespace-id"), ctx.Query("namespace"))
	request = ControllerInvokeRequest{
		RequestID:   strings.TrimSpace(ctx.RequestID),
		TenantID:    strings.TrimSpace(ctx.TenantID),
		Namespace:   namespace,
		AuthContext: AuthContext{Subject: strings.TrimSpace(ctx.AuthIdentity), TenantID: strings.TrimSpace(ctx.TenantID), Namespace: namespace},
		SessionID:   sessionID,
		Method:      ctx.Request.Method,
		Path:        proxyPath,
		Query:       cloneInvokeQueryValues(ctx.Request.Query),
		Headers:     sanitizeProviderInvokeHeaders(ctx.Request.Headers),
		Body:        append([]byte(nil), ctx.Request.Body...),
		Port:        firstPositiveInt32(parseInt32(ctx.Header("x-apptheory-microvm-port")), defaultProviderInvokePort),
		TTLSeconds:  firstPositiveInt32(parseInt32(ctx.Header("x-apptheory-microvm-token-ttl")), defaultProviderInvokeTTLSeconds),
	}
	return normalizeControllerInvokeRequest(request), SafeError{}
}

func controllerHTTPResponse(response ControllerResponse) *apptheory.Response {
	status := controllerHTTPStatus(response.Error)
	out, err := apptheory.JSON(status, response)
	if err != nil {
		return apptheory.Text(500, "internal error")
	}
	return out
}

func controllerInvokeHTTPResponse(output ProviderInvokeOutput) *apptheory.Response {
	status := output.Status
	if status == 0 {
		status = 502
	}
	return &apptheory.Response{
		Status:   status,
		Headers:  sanitizeProviderInvokeHeaders(output.Headers),
		Body:     append([]byte(nil), output.Body...),
		IsBase64: output.IsBase64,
	}
}

func controllerInvokeMethods() []string {
	return []string{"DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"}
}

func cloneInvokeQueryValues(query map[string][]string) map[string][]string {
	out := cloneQueryValues(query)
	delete(out, "tenant_id")
	delete(out, "namespace")
	return out
}

func controllerHTTPStatus(err *SafeError) int {
	if err == nil || err.Code == "" {
		return 200
	}
	switch err.Code {
	case ErrorCodeUnauthenticatedController:
		return 401
	case ErrorCodeTenantBindingViolation:
		return 403
	case ErrorCodeSessionRegistryIncomplete:
		return 404
	case ErrorCodeControllerIncomplete:
		return 500
	case ErrorCodeControllerCommandFailed, ErrorCodeProviderOperationFailed:
		return 502
	default:
		return 400
	}
}

func parseInt32(value string) int32 {
	n, err := strconv.ParseInt(strings.TrimSpace(value), 10, 32)
	if err != nil || n <= 0 {
		return 0
	}
	return int32(n)
}

func firstPositiveInt32(value int32, fallback int32) int32 {
	if value > 0 {
		return value
	}
	return fallback
}

File: ts/src/microvm/controller-routes.ts
import type { App } from "../app.js";
import type { Context } from "../context.js";
import { json as jsonResponse } from "../response.js";
import type { Headers, Query, Response } from "../types.js";

import {
  MICROVM_ERROR_CONTROLLER_COMMAND_FAILED,
  MICROVM_ERROR_CONTROLLER_INCOMPLETE,
  MICROVM_ERROR_INVALID_CONTROLLER_REQUEST,
  MICROVM_ERROR_PROVIDER_OPERATION_FAILED,
  MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
  MICROVM_ERROR_TENANT_BINDING_VIOLATION,
  MICROVM_ERROR_UNAUTHENTICATED_CONTROLLER,
  MicroVMCommand,
  MicroVMSafeError,
  type MicroVMCommandName,
  type MicroVMControllerRequest,
  type MicroVMControllerResponse,
  type MicroVMControllerRouteTarget,
  type MicroVMProviderIdlePolicy,
  type MicroVMProviderPortScope,
  type MicroVMSessionSpec,
} from "./model.js";
import { safeError } from "./errors.js";
import {
  controllerErrorResponse,
  normalizeMicroVMControllerRequest,
} from "./controller.js";
import { normalizeMicroVMCommand } from "./controller-contract.js";
import { normalizeStringArray } from "./provider.js";

export function registerMicroVMControllerRoutes(
  app: App,
  controller: MicroVMControllerRouteTarget,
): App {
  if (!app) {
    throw new Error(
      "apptheory: microvm controller route registration requires an app",
    );
  }
  if (!controller) {
    throw safeError(
      MICROVM_ERROR_CONTROLLER_INCOMPLETE,
      "apptheory: microvm controller route registration requires a controller",
      "",
    );
  }
  const routes: Array<{
    method: string;
    path: string;
    command: MicroVMCommandName;
  }> = [
    { method: "POST", path: "/microvms", command: MicroVMCommand.Run },
    { method: "GET", path: "/microvms", command: MicroVMCommand.List },
    {
      method: "GET",
      path: "/microvms/{session_id}",
      command: MicroVMCommand.Get,
    },
    {
      method: "POST",
      path: "/microvms/{session_id}/suspend",
      command: MicroVMCommand.Suspend,
    },
    {
      method: "POST",
      path: "/microvms/{session_id}/resume",
      command: MicroVMCommand.Resume,
    },
    {
      method: "DELETE",
      path: "/microvms/{session_id}",
      command: MicroVMCommand.Terminate,
    },
    {
      method: "POST",
      path: "/microvms/{session_id}/auth-token",
      command: MicroVMCommand.AuthToken,
    },
    {
      method: "POST",
      path: "/microvms/{session_id}/shell-auth-token",
      command: MicroVMCommand.ShellAuthToken,
    },
    {
      method: "POST",
      path: "/microvms/{session_id}/shell-token",
      command: MicroVMCommand.ShellAuthToken,
    },
  ];
  for (const route of routes) {
    app.handleStrict(
      route.method,
      route.path,
      microVMControllerRouteHandler(controller, route.command),
      { authRequired: true },
    );
  }
  return app;
}

export function registerControllerRoutes(
  app: App,
  controller: MicroVMControllerRouteTarget,
): App {
  return registerMicroVMControllerRoutes(app, controller);
}

export function microVMControllerRouteHandler(
  controller: MicroVMControllerRouteTarget,
  command: MicroVMCommandName,
): (ctx: Context) => Promise<Response> {
  return async (ctx: Context): Promise<Response> => {
    const parsed = microVMControllerRequestFromHTTP(ctx, command);
    if (parsed instanceof MicroVMSafeError) {
      const request: MicroVMControllerRequest = {
        command,
        request_id: String(ctx?.requestId ?? "").trim(),
        tenant_id: String(ctx?.tenantId ?? "").trim(),
        namespace: "",
        auth_context: {
          subject: String(ctx?.authIdentity ?? "").trim(),
          tenant_id: String(ctx?.tenantId ?? "").trim(),
        },
      };
      return microVMControllerHTTPResponse(
        controllerErrorResponse(request, parsed),
      );
    }
    const response = await controller.handle(parsed);
    return microVMControllerHTTPResponse(response);
  };
}

export function microVMControllerRequestFromHTTP(
  ctx: Context,
  command: MicroVMCommandName,
): MicroVMControllerRequest | MicroVMSafeError {
  if (!ctx) {
    return safeError(
      MICROVM_ERROR_INVALID_CONTROLLER_REQUEST,
      "apptheory: microvm controller route context is missing",
      "",
    );
  }
  const payloadResult = microVMControllerRoutePayload(ctx);
  if (payloadResult instanceof MicroVMSafeError) return payloadResult;
  const payload = payloadResult;
  const pathSessionID = String(ctx.param("session_id") ?? "").trim();
  const bodySessionID = stringFromPayload(payload, "session_id");
  if (pathSessionID && bodySessionID && pathSessionID !== bodySessionID) {
    return safeError(
      MICROVM_ERROR_TENANT_BINDING_VIOLATION,
      "apptheory: microvm controller route session binding mismatch",
      ctx.requestId,
    );
  }
  const ctxTenant = String(ctx.tenantId ?? "").trim();
  const bodyTenant = stringFromPayload(payload, "tenant_id");
  const queryTenant = firstQueryValue(ctx.request.query, "tenant_id");
  if (ctxTenant && bodyTenant && bodyTenant !== ctxTenant) {
    return safeError(
      MICROVM_ERROR_TENANT_BINDING_VIOLATION,
      "apptheory: microvm controller route tenant binding mismatch",
      ctx.requestId,
    );
  }
  if (ctxTenant && queryTenant && queryTenant !== ctxTenant) {
    return safeError(
      MICROVM_ERROR_TENANT_BINDING_VIOLATION,
      "apptheory: microvm controller route tenant binding mismatch",
      ctx.requestId,
    );
  }
  const namespace =
    stringFromPayload(payload, "namespace") ||
    firstHeaderValueFromMap(ctx.request.headers, "x-namespace-id") ||
    firstQueryValue(ctx.request.query, "namespace");
  const request: MicroVMControllerRequest = {
    command: normalizeMicroVMCommand(command),
    request_id: String(ctx.requestId ?? "").trim(),
    tenant_id: ctxTenant || bodyTenant || queryTenant,
    namespace,
    auth_context: {
      subject: String(ctx.authIdentity ?? "").trim(),
      tenant_id: ctxTenant,
      namespace,
    },
    session_id: pathSessionID || bodySessionID,
    image_ref: stringFromPayload(payload, "image_ref"),
    image_version: stringFromPayload(payload, "image_version"),
    network_connector_ref: stringFromPayload(payload, "network_connector_ref"),
    ingress_network_connector_refs: stringListFromPayload(
      payload,
      "ingress_network_connector_refs",
    ),
    egress_network_connector_refs: stringListFromPayload(
      payload,
      "egress_network_connector_refs",
    ),
    session_spec: sessionSpecFromPayload(payload),
    maximum_duration_seconds: intFromPayload(
      payload,
      "maximum_duration_seconds",
    ),
    ttl_seconds: intFromPayload(payload, "ttl_seconds"),
    allowed_port_scope: portScopesFromPayload(payload),
    max_results:
      positiveIntFromPayload(payload, "max_results") ||
      positiveIntFromString(firstQueryValue(ctx.request.query, "max_results")),
  };
  const idlePolicy = idlePolicyFromPayload(payload);
  if (idlePolicy) request.idle_policy = idlePolicy;
  return normalizeMicroVMControllerRequest(request);
}

export function microVMControllerRoutePayload(
  ctx: Context,
): Record<string, unknown> | MicroVMSafeError {
  if ((ctx.request.body?.length ?? 0) === 0) return {};
  try {
    const parsed: unknown = JSON.parse(
      Buffer.from(ctx.request.body).toString("utf8"),
    );
    if (
      parsed === null ||
      typeof parsed !== "object" ||
      Array.isArray(parsed)
    ) {
      return safeError(
        MICROVM_ERROR_INVALID_CONTROLLER_REQUEST,
        "apptheory: microvm controller route request is malformed",
        ctx.requestId,
      );
    }
    return parsed as Record<string, unknown>;
  } catch {
    return safeError(
      MICROVM_ERROR_INVALID_CONTROLLER_REQUEST,
      "apptheory: microvm controller route request is malformed",
      ctx.requestId,
    );
  }
}

export function microVMControllerHTTPResponse(
  response: MicroVMControllerResponse,
): Response {
  return jsonResponse(
    microVMControllerHTTPStatus(response.error),
    serializableMicroVMControllerResponse(response),
  );
}

export function microVMControllerHTTPStatus(err?: MicroVMSafeError): number {
  if (!err || !err.code) return 200;
  switch (err.code) {
    case MICROVM_ERROR_UNAUTHENTICATED_CONTROLLER:
      return 401;
    case MICROVM_ERROR_TENANT_BINDING_VIOLATION:
      return 403;
    case MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE:
      return 404;
    case MICROVM_ERROR_CONTROLLER_INCOMPLETE:
      return 500;
    case MICROVM_ERROR_CONTROLLER_COMMAND_FAILED:
    case MICROVM_ERROR_PROVIDER_OPERATION_FAILED:
      return 502;
    default:
      return 400;
  }
}

export function serializableMicroVMControllerResponse(
  response: MicroVMControllerResponse,
): Record<string, unknown> {
  const out: Record<string, unknown> = { ...response };
  if (response.error) {
    out["error"] = {
      code: response.error.code,
      message: response.error.message,
      request_id: response.error.request_id ?? "",
    };
  }
  return out;
}

export function firstHeaderValueFromMap(
  headers: Headers,
  name: string,
): string {
  const key = String(name ?? "")
    .trim()
    .toLowerCase();
  const values = headers[key] ?? headers[String(name ?? "").trim()] ?? [];
  return String(values[0] ?? "").trim();
}

export function firstQueryValue(query: Query, name: string): string {
  const values = query[String(name ?? "").trim()] ?? [];
  return String(values[0] ?? "").trim();
}

export function stringFromPayload(
  payload: Record<string, unknown>,
  key: string,
): string {
  return String(payload[key] ?? "").trim();
}

export function stringListFromPayload(
  payload: Record<string, unknown>,
  key: string,
): string[] {
  const value = payload[key];
  if (!Array.isArray(value)) return [];
  return normalizeStringArray(value.map((item) => String(item ?? "")));
}

export function intFromPayload(
  payload: Record<string, unknown>,
  key: string,
): number {
  return Math.trunc(Number(payload[key] ?? 0) || 0);
}

export function positiveIntFromPayload(
  payload: Record<string, unknown>,
  key: string,
): number {
  const value = intFromPayload(payload, key);
  return value > 0 ? value : 0;
}

export function positiveIntFromString(value: string): number {
  const parsed = Math.trunc(Number(String(value ?? "").trim()) || 0);
  return parsed > 0 ? parsed : 0;
}

export function sessionSpecFromPayload(
  payload: Record<string, unknown>,
): MicroVMSessionSpec {
  const raw = payload["session_spec"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return {};
  const metadata = (raw as Record<string, unknown>)["metadata"];
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    return {};
  }
  const out: Record<string, string> = {};
  for (const [key, value] of Object.entries(metadata)) {
    out[String(key).trim()] = String(value ?? "");
  }
  return Object.keys(out).length > 0 ? { metadata: out } : {};
}

export function idlePolicyFromPayload(
  payload: Record<string, unknown>,
): MicroVMProviderIdlePolicy | undefined {
  const raw = payload["idle_policy"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return undefined;
  const record = raw as Record<string, unknown>;
  return {
    auto_resume_enabled: record["auto_resume_enabled"] === true,
    max_idle_duration_seconds: Math.trunc(
      Number(record["max_idle_duration_seconds"] ?? 0) || 0,
    ),
    suspended_duration_seconds: Math.trunc(
      Number(record["suspended_duration_seconds"] ?? 0) || 0,
    ),
  };
}

export function portScopesFromPayload(
  payload: Record<string, unknown>,
): MicroVMProviderPortScope[] {
  const raw = payload["allowed_port_scope"];
  if (!Array.isArray(raw)) return [];
  return raw
    .filter(
      (item): item is Record<string, unknown> =>
        Boolean(item) && typeof item === "object" && !Array.isArray(item),
    )
    .map((scope) => ({
      all_ports: scope["all_ports"] === true,
      port: Math.trunc(Number(scope["port"] ?? 0) || 0),
      start_port: Math.trunc(Number(scope["start_port"] ?? 0) || 0),
      end_port: Math.trunc(Number(scope["end_port"] ?? 0) || 0),
    }));
}
